### PR TITLE
Added type definitions for Parcel bundler

### PR DESCRIPTION
**Pull request by pinage404**
Using this page https://en.parceljs.org/api.html

Tests force me to export everything, i don't know the good way to make it work without exporting everything

### CODE REVIEW

File: types/parcel-bundler/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+     * A Map<String(Type: js, css, map, ...), Bundle> of all sibling bundles
+     */
+    siblingBundlesMap: Map<string, ParcelBundle>;
+    /**
+     * A Map<Asset, number(line number inside the bundle)> of all the locations of the assets inside the bundle, used to generate accurate source maps
+     */
+    offsets: Map<ParcelAsset, number>;
+}
+
+export default class ParcelBundler {
```
Comment (weswigham) [inline types/parcel-bundler/index.d.ts]: This [should be](https://github.com/parcel-bundler/parcel/blob/master/src/Bundler.js#L788)
```ts
class ParcelBundler {
    constructor(entryFiles?: string | string[], options?: ParcelOptions);
    bundle(): Promise<ParcelBundle>;
}
exports = ParcelBundler;
```
and if your runtime module loader makes a cjs module available as a default, use the `esModuleInterop` and `allowSyntheticDefaultImports` compiler options.

**Review by weswigham — CHANGES_REQUESTED** [review]

File: types/parcel-bundler/index.d.ts (hunk 1 above)
Comment (pinage404) [reply types/parcel-bundler/index.d.ts]: Hi,

I tried and i got this errors when running `npm run test`

```
Error: ./types/parcel-bundler/index.d.ts:163:1
ERROR: 163:1  expect                           TypeScript@next compile error:
A 'declare' modifier is required for a top level declaration in a .d.ts file.
ERROR: 163:7  strict-export-declare-modifiers  Prefer to explicitly write 'export' for an external module export. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 169:1  expect                           TypeScript@next compile error:
Statements are not allowed in ambient contexts.
ERROR: 169:1  expect                           TypeScript@next compile error:
Cannot find name 'exports'.
ERROR: 169:1  strict-export-declare-modifiers  Prefer to explicitly write 'export' for an external module export. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md

./types/parcel-bundler/parcel-bundler-tests.ts:1:8
ERROR: 1:8    expect                           TypeScript@next compile error:
Module '"./types/parcel-bundler/index"' has no default export.
ERROR: 9:23   expect                           TypeScript@next compile error:
Parameter 'bundle' implicitly has an 'any' type.
```

So, i tried to add a `declare`

```diff
-class ParcelBundler {
+declare class ParcelBundler {
     constructor(entryFiles?: string | string[], options?: ParcelOptions);
 
     bundle(): Promise<ParcelBundle>;
 }
 
 exports = ParcelBundler;
```
And i got this errors

```
Error in parcel-bundler
Error: ./types/parcel-bundler/index.d.ts:163:1
ERROR: 163:1  strict-export-declare-modifiers  Prefer 'export' to 'declare' in an external module. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 169:1  expect                           TypeScript@next compile error:
Statements are not allowed in ambient contexts.
ERROR: 169:1  expect                           TypeScript@next compile error:
Cannot find name 'exports'.
ERROR: 169:1  strict-export-declare-modifiers  Prefer to explicitly write 'export' for an external module export. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md

./types/parcel-bundler/parcel-bundler-tests.ts:1:8
ERROR: 1:8    expect                           TypeScript@next compile error:
Module '"./types/parcel-bundler/index"' has no default export.
ERROR: 9:23   expect                           TypeScript@next compile error:
Parameter 'bundle' implicitly has an 'any' type.
```

I don't understand what i did wrong, i'm new in the TypeScript ecosystem, so i don't understand everything

What should i do ?

Comment (weswigham) [reply types/parcel-bundler/index.d.ts]: Ah, my fault. It's `export =` not `exports =`. I typo'd.

Comment (pinage404) [reply types/parcel-bundler/index.d.ts]: i tried this

```typescript
declare class ParcelBundler {
    constructor(entryFiles?: string | string[], options?: ParcelOptions);

    bundle(): Promise<ParcelBundle>;
}

export = ParcelBundler;
```

and got this

```
Error in parcel-bundler
Error: ./types/parcel-bundler/index.d.ts:169:1
ERROR: 169:1  expect  TypeScript@next compile error:
An export assignment cannot be used in a module with other exported elements.

./types/parcel-bundler/parcel-bundler-tests.ts:1:8
ERROR: 1:8    expect  TypeScript@next compile error:
Module '"./types/parcel-bundler/index"' has no default export.
ERROR: 1:25   expect  TypeScript@next compile error:
Module '"./types/parcel-bundler/index"' has no exported member 'ParcelOptions'.
ERROR: 1:46   expect  TypeScript@next compile error:
Module '"./types/parcel-bundler/index"' resolves to a non-module entity and cannot be imported using thisconstruct.
ERROR: 9:23   expect  TypeScript@next compile error:
Parameter 'bundle' implicitly has an 'any' type.
```

Comment (weswigham) [reply types/parcel-bundler/index.d.ts]: Righto, you'll need to wrap the other elements in a `declare namespace ParcelBundler` declaration so they all become elements under the export assigned class/namespace merge.

Comment (pinage404) [reply types/parcel-bundler/index.d.ts]: i tried and this didn't work, you can see my last commits

**Comment by pinage404** [comment]
i tried outside, but the compiler can't find namespace members

**Comment by weswigham** [comment]
Right, wthin the file but outside of the namespace, they need to be qualified, ie `ParcelBundler.ParcelBundle`

**Comment by pinage404** [comment]
Thanks ! It works !

I had to add `"allowSyntheticDefaultImports": true` to make the test pass

**Comment by pinage404** [comment]
Can you explain to me why it's better than the first solution ? What is different ?

**Comment by weswigham** [comment]
The cjs module parcel exports actually sets the entire export object to the class - it's only available as a default under certain (loader) configurations. This captures that, and allows, eg `const Parcel = require("parcel-bundler")` to work for JS consumers.

**Pull request merged**